src/wp-includes/post.php

```diff
@@ -3400,21 +3400,45 @@ function wp_count_posts( $type = 'post', $perm = '' ) {
 		return apply_filters( 'wp_count_posts', $counts, $type, $perm );
 	}
 
-	$query = "SELECT post_status, COUNT( * ) AS num_posts FROM {$wpdb->posts} WHERE post_type = %s";
+	$query = $wpdb->prepare( "SELECT post_status, COUNT(*) AS num_posts FROM {$wpdb->posts} WHERE post_type = %s", $type );
 
 	if ( 'readable' === $perm && is_user_logged_in() ) {
 		$post_type_object = get_post_type_object( $type );
 		if ( ! current_user_can( $post_type_object->cap->read_private_posts ) ) {
```

Comment:
These conditions can be combined, and then the `$query` above can be set in the `else`.

```suggestion
	if ( 'readable' === $perm && is_user_logged_in() && ! current_user_can( get_post_type_object( $type )->cap->read_private_posts ) ) {
```